### ISSUE TITLE
Can't compile on Linux

### ISSUE THREAD

**Issue by NewNodeGames** [BastiaanOlij/gdprocmesh#22]
First steps work well

cd godot-cpp
scons platform=x11 generate_bindings=yes use_custom_api_file=yes custom_api_file=../api.json
cd ..

but on

scons platform=x11

this is the backtrace:

scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
g++ -o src/gdprocmesh.os -c -fPIC -I. -Isrc -Igodot-cpp/godot_headers -Igodot-cpp/include -Igodot-cpp/include/core -Igodot-cpp/include/gen src/gdprocmesh.cpp
src/gdprocmesh.cpp: In member function 'void godot::GDProcMesh::_update()':
src/gdprocmesh.cpp:645:64: error: call of overloaded 'PoolVector3Array(godot::Variant&)' is ambiguous
      } else if (((PoolVector3Array) arr[ArrayMesh::ARRAY_VERTEX]).size() == 0) {
                                                                ^
In file included from godot-cpp/include/core/Variant.hpp:13,
                 from godot-cpp/include/core/Dictionary.hpp:4,
                 from godot-cpp/include/core/CoreTypes.hpp:10,
                 from godot-cpp/include/core/Godot.hpp:11,
                 from src/gdprocmesh.h:4,
                 from src/gdprocmesh.cpp:1:
godot-cpp/include/core/PoolArrays.hpp:566:2: note: candidate: 'godot::PoolVector3Array::PoolVector3Array(const godot::Array&)'
  PoolVector3Array(const Array& array);
  ^~~~~~~~~~~~~~~~
godot-cpp/include/core/PoolArrays.hpp:563:2: note: candidate: 'godot::PoolVector3Array::PoolVector3Array(const godot::PoolVector3Array&)'
  PoolVector3Array(const PoolVector3Array &p_other);
  ^~~~~~~~~~~~~~~~
src/gdprocmesh.cpp:649:59: error: call of overloaded 'PoolIntArray(godot::Variant&)' is ambiguous
      } else if (((PoolIntArray) arr[ArrayMesh::ARRAY_INDEX]).size() == 0) {
                                                           ^
In file included from godot-cpp/include/core/Variant.hpp:13,
                 from godot-cpp/include/core/Dictionary.hpp:4,
                 from godot-cpp/include/core/CoreTypes.hpp:10,
                 from godot-cpp/include/core/Godot.hpp:11,
                 from src/gdprocmesh.h:4,
                 from src/gdprocmesh.cpp:1:
godot-cpp/include/core/PoolArrays.hpp:181:2: note: candidate: 'godot::PoolIntArray::PoolIntArray(const godot::Array&)'
  PoolIntArray(const Array& array);
  ^~~~~~~~~~~~
godot-cpp/include/core/PoolArrays.hpp:178:2: note: candidate: 'godot::PoolIntArray::PoolIntArray(const godot::PoolIntArray&)'
  PoolIntArray(const PoolIntArray &p_other);
  ^~~~~~~~~~~~
scons: *** [src/gdprocmesh.os] Error 1
scons: building terminated because of errors.

**Comment by NewNodeGames** [comment]
I tried to replace "x11" for "linux" and results to a bigger bin file but compilation fails too.

**Comment by BastiaanOlij** [comment]
Yes for both godot-cpp and gdprocmesh the scons file looks for platform=linux
Not sure why it would fail on compiling, I work mostly on Windows.
@capnm you got it to compile on Linux in the end didn't you? Any ideas what might be causing this?

**Comment by BastiaanOlij** [comment]
@capnm for some reason godot-cpp went for linux so I stuck with that. Could have been an accidental fault from an earlier scons file either me or Karroffel adopted.

I think checking for either makes sense, I also need to add an else so I can just error out if you specify a platform that is unknown to scons

**Comment by BastiaanOlij** [comment]
I've done the platform check improvements

**Comment by capnm** [comment]
You can use the built-in  `EnumVariable`, for example:
```python
opts.Add(EnumVariable('b',
        "Set the build type or the clang address/memory sanitizer.", 'd',
        ['d','debug', 'p', 'production', 'asan', 'msan']
))
```

**Comment by QbieShay** [comment]
Not working on my machine. Tried with branch nativescript1.1, tried with master, it fails in both cases.

**Comment by w84death** [comment]
I was going to write the same bugreport.

`src/gdprocmesh.cpp: In member function 'void godot::GDProcMesh::_update()':
src/gdprocmesh.cpp:646:63: error: call of overloaded 'PoolVector3Array(godot::Variant&)' is ambiguous
      } else if (((PoolVector3Array)arr[ArrayMesh::ARRAY_VERTEX]).size() == 0) {
                                                               ^
`

Linux, Debian (Ubuntu), 64bit.

**Comment by BastiaanOlij** [comment]
Very strange, so the compiler can't tell types apart. I guess we can write out that part of code into multiple steps but unless the compiler doesn't handle the newer C++ spec well it shouldn't fail on that..

**Comment by BastiaanOlij** [comment]
When I next run into Karroffel need to have a chat with him, he knows this stuff alot better then I do. This all works with the MSVCs compiler I use so, I don't do much on Linux I'm afraid.

You can try writing out the bit of code that it fails on. The newer compilers are better at deciding which function to call if there are multiple functions that could potentially match the variable types. 

So its a matter of making sure you tell it what types are being used. So
```
if (((PoolVector3Array)arr[ArrayMesh::ARRAY_VERTEX]).size() == 0)
```

would become
```
Variant v = arr[ArrayMesh::Array_VERTEX];
PoolVector3Array pv = v;
if (pv.size() == 0)
```

Or something along those lines.

**Comment by TheSHEEEP** [comment]
Just chiming in on this, with the following commands I can build this module just fine:
`scons -j4 platform=linux bits=64 target=release generate_bindings=yes use_custom_api_file=yes custom_api_file=../api.json`
for godot-cpp

and
`scons -j4 latform=linux bits=64 target=release`
for the module itself

Unfortunately, it doesn't quite work, but I'll open a new issue for that.

**Comment by BastiaanOlij** [comment]
Let me know how it goes with the changes I just made. Note that the custom api.json should no longer be needed
